File: fabric-audit-agent/core/detectors/cost.js
import { DEFAULT_CONFIG } from '../config.js';

/** @typedef {import('./capacity.js').Flag} Flag */

/**
 * Cost / unused-resource detectors. Pure: facts in, flags out.
 * @param {{usage?:object}} facts
 * @param {object} [config]
 * @returns {Flag[]}
 */
export function detectCost(facts, config = DEFAULT_CONFIG) {
  const u = facts?.usage ?? {};
  const flags = [];

  for (const r of u.reports ?? []) {
    if ((r.views30d ?? 0) === 0) {
      flags.push({
        type: 'cost.unused-report',
        resource: `${r.workspace} / ${r.name}`,
        when: '',
        evidence: { views30d: 0 },
        what: `Report "${r.name}" has had 0 views in 30 days.`,
      });
    }
  }

  for (const c of u.capacities ?? []) {
    if ((c.avgCuPct ?? 100) < config.cost.idleCuPct) {
      flags.push({
        type: 'cost.idle-capacity',
        resource: `capacity ${c.id}`,
        when: '',
        evidence: { sku: c.sku, avgCuPct: c.avgCuPct },
        what: `Capacity ${c.id} (${c.sku}) averaged ${c.avgCuPct}% CU — largely idle.`,
      });
    }
  }

  return flags;
}

File: fabric-audit-agent-app/fabric_audit_agent/detectors/cost.py
"""Cost / unused-resource detectors. Faithful port of the Node ``core/detectors/cost.js``."""
from ..config import DEFAULT_CONFIG


def detect_cost(facts, config=None):
    config = config or DEFAULT_CONFIG
    u = (facts or {}).get("usage") or {}
    cost = config["cost"]
    flags = []

    for r in (u.get("reports") or []):
        # Only flag CONFIRMED zero — not missing data. The old `(views30d or 0) == 0` treated
        # both a genuine 0 view count and a None (no data collected) as "unused," flagging reports
        # we simply lack telemetry for as if they were confirmed orphans (2026-08-09 audit).
        if r.get("views30d") == 0:
            flags.append({
                "type": "cost.unused-report",
                "resource": f"{r.get('workspace')} / {r.get('name')}",
                "when": "",
                "evidence": {"views30d": 0},
                "what": f"Report \"{r.get('name')}\" has had 0 views in 30 days.",
            })

    for c in (u.get("capacities") or []):
        avg = c.get("avgCuPct")
        avg = avg if avg is not None else 100   # JS: c.avgCuPct ?? 100 (nullish, so 0 stays 0)
        if avg < cost["idleCuPct"]:
            flags.append({
                "type": "cost.idle-capacity",
                "resource": f"capacity {c.get('id')}",
                "when": "",
                "evidence": {"sku": c.get("sku"), "avgCuPct": c.get("avgCuPct")},
                "what": f"Capacity {c.get('id')} ({c.get('sku')}) averaged {c.get('avgCuPct')}% CU — largely idle.",
            })
    return flags
